genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+import com.netflix.genie.agent.execution.services.FetchingCacheService;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+import org.springframework.util.DigestUtils;
+import org.springframework.util.FileCopyUtils;
+import org.springframework.util.FileSystemUtils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.RandomAccessFile;
+import java.net.URI;
+import java.nio.channels.FileChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * A file system based cache on local disk that uses URIs as keys and transparently downloads
+ * missing entries. It uses OS level locks to accomplish the following
+ * <p>
+ * Allows multiple agents reading/writing/deleting in the same cache folder
+ * without corrupting the cache
+ * Downloads new versions of resources once they are available remotely
+ * Deletes the old versions of resources once a new version is successfully downloaded
+ * Recovers from partial downloads of resources in case an agent gets killed in the middle of a download
+ * or a download fails for any other reason
+ * <p>
+ * Does NOT handle concurrency within the same agent (not an issue at the moment)
+ * <p>
+ * Cache structure on local disk
+ * Each resource has a hash to represent it. The version of the resource is the remote last modified
+ * timestamp. For each resource and version, a lock file is created. Each process takes a lock on
+ * this file and then downloads the resource to a data.tmp file. Once the download is complete
+ * data.tmp is atomically moved to data.
+ * For e.g for a hash value of 6d331abc92bc8244bc5d41e2107f303a and last modified = 1525456404
+ * cache entries would like the following
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/data
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/lock
+ * <p>
+ * Deletion of older versions
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Service
+@Lazy
+@Slf4j
+@Component
+class FileSystemFetchingCacheService implements FetchingCacheService {
+
+    public static final String LOCK_FILE_NAME = "lock";
+    public static final String DATA_FILE_NAME = "data";
+    public static final String DATA_DOWNLOAD_FILE_NAME = "data.tmp";
+    protected final ResourceLoader resourceLoader;
+    protected final File cacheDirectory;
+    protected final LockingFactory lockingFactory;
+    private final ExecutorService oldResourcesCleanUpService = Executors.newFixedThreadPool(1);
```

Comment:
Not sure I love the idea of this "hidden" thread pool sitting inside this class. Unless we look into all the implementations we'd have no idea it existed. It's probably something that should be provided to the class via the constructor (easier for testing, debugging, configuring, etc.). Also you should configure the thread pool so the names of the threads are clear. In this case they'd just show up in the log very generically.

Reply:
I was intentionally trying to restrict the logic of how we do the clean up - asynchronously or synchronously. Hence did not expose it.

Reply:
We are unlikely to ever have enough time to go back and fix a JIRA about injecting an executor. Higher priority things will always come first.
Moreover if we ever get to it, whoever takes it, needs to re-understand this code, which is not very efficient for such a minor change.
So we either keep this as is and let's not create a JIRA, or we fix this right now and commit complete code that works and whose quality we're all happy with.